packages/expo-modules-core/ios/JSI/JavaScriptObject.mm

```diff
@@ -0,0 +1,71 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
+
+#import <ExpoModulesCore/JSIConversions.h>
+#import <ExpoModulesCore/JavaScriptObject.h>
+#import <ExpoModulesCore/JavaScriptRuntime.h>
+#import <ExpoModulesCore/ExpoModulesProxySpec.h>
+
+@implementation JavaScriptObject {
+  __weak JavaScriptRuntime *_runtime;
+  std::shared_ptr<jsi::Object> _jsObjectPtr;
+}
+
+- (nonnull instancetype)initWith:(std::shared_ptr<jsi::Object>)jsObject
+                         runtime:(nonnull JavaScriptRuntime *)runtime
+{
+  if (self = [super init]) {
+    _runtime = runtime;
+    _jsObjectPtr = jsObject;
+  }
+  return self;
+}
+
+- (jsi::Object *)get
+{
+  return _jsObjectPtr.get();
+}
+
+#pragma mark - Subscripting
+
+- (nullable id)objectForKeyedSubscript:(nonnull NSString *)key
+{
+  auto runtime = [_runtime get];
+  auto callInvoker = [_runtime callInvoker];
+
+  if (runtime && callInvoker) {
+    auto value = _jsObjectPtr->getProperty(*runtime, [key UTF8String]);
+    return expo::convertJSIValueToObjCObject(*runtime, value, callInvoker);
+  }
+  return nil;
+}
+
+- (void)setObject:(nullable id)obj forKeyedSubscript:(nonnull NSString *)key
+{
+  if (auto runtime = [_runtime get]) {
+    if ([obj isKindOfClass:[JavaScriptObject class]]) {
+      _jsObjectPtr->setProperty(*runtime, [key UTF8String], *[obj get]);
+    } else {
+      _jsObjectPtr->setProperty(*runtime, [key UTF8String], expo::convertObjCObjectToJSIValue(*runtime, obj));
+    }
+  }
+}
+
+#pragma mark - Functions
+
+- (void)setAsyncFunction:(nonnull NSString *)name
+               argsCount:(NSInteger)argsCount
+                   block:(nonnull JSAsyncFunctionBlock)block
+{
+  jsi::Function function = [_runtime createAsyncFunction:name argsCount:argsCount block:block];
+  _jsObjectPtr->setProperty(*[_runtime get], [name UTF8String], function);
```

Comment:
I shoulda check if runtime is not null

Reply:
Do you? You marked runtime as `nonnull`able above in the same file 🤔

Reply:
It's `nonnull` only when it's passed as an argument to the initializer of `JavaScriptObject`. `_runtime` is marked as a `__weak` property, so it might be deallocated before the object gets deallocated.